File: solutions/golang/1281/solution.go
var numberOfPeople int
func maximizeSweetness(sweetness []int, k int) int {
    numberOfPeople = k + 1
    left := math.MaxInt
    right := 0
    // Get minimum element and the sum of all sweetness
    for _, num := range sweetness {
        if left > num {
            left = num
        }
        right += num
    }

    right /= numberOfPeople
    for left <= right {
        mid := left + (right - left) / 2
        if check(mid, sweetness) {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    return right
}

func check(minSweetness int, sweetness []int) bool {
    peopleWithChocolate := 0
    currSweetness := 0
    for _, s := range sweetness {
        currSweetness += s
        if currSweetness >= minSweetness {
            currSweetness = 0
            peopleWithChocolate++
        }
    }

    return peopleWithChocolate >= numberOfPeople
}

File: solutions/cpp/1281/solution.cpp
class Solution {
public:
    int numberofPeople;
    int maximizeSweetness(vector<int>& sweetness, int k) {
        numberofPeople = k + 1;
        int left = *min_element(sweetness.begin(), sweetness.end());
        int right = accumulate(sweetness.begin(), sweetness.end(), 0) / numberofPeople;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (check(mid, sweetness)) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return right;
    }

    bool check(int minSweetness, vector<int> &sweetness) {
        int peopleWithChocolate = 0;
        int currSweetness = 0;
        for (int s : sweetness) {
            currSweetness += s;
            if (currSweetness >= minSweetness) {
                peopleWithChocolate++;
                currSweetness = 0;
            }
        }

        return peopleWithChocolate >= numberofPeople;
    }
};
